shiny/ui/_theme_brand.py

```diff
@@ -0,0 +1,410 @@
+        sass_vars_colors: dict[str, str] = {}
+        sass_vars_brand_colors: dict[str, str] = {}
+        css_vars_brand_colors: list[str] = []
+
+        raise_for_unmapped_vars = (
+            os.environ.get("SHINY_BRAND_YML_RAISE_UNMAPPED") == "true"
+        )
+
+        if brand.color:
+            # Map values in colors to their Sass variable counterparts
+            for thm_name, thm_color in brand.color.to_dict(include="theme").items():
+                if thm_name not in color_map:
+                    if raise_for_unmapped_vars:
+                        raise ThemeBrandUnmappedFieldError(f"color.{thm_name}")
+                    continue
+
+                for sass_var in color_map[thm_name]:
+                    sass_vars_colors[sass_var] = thm_color
+
+            brand_color_palette = brand.color.to_dict(include="palette")
+
+            # Map the brand color palette to Bootstrap's named colors, e.g. $red, $blue.
+            for pal_name, pal_color in brand_color_palette.items():
+                if pal_name in bootstrap_colors:
+                    sass_vars_colors[pal_name] = pal_color
+
+                # Create Sass and CSS variables for the brand color palette
+                color_var = sanitize_sass_var_name(pal_name)
+
+                # => Sass var: `$brand-{name}: {value}`
+                sass_vars_brand_colors.update({f"brand-{color_var}": pal_color})
+                # => CSS var: `--brand-{name}: {value}`
+                css_vars_brand_colors.append(f"--brand-{color_var}: {pal_color};")
+
+        # brand.typography ------------------------------------------------------------
+        sass_vars_typography: dict[str, str] = {}
+        if brand.typography:
+            brand_typography = brand.typography.model_dump(
+                exclude={"fonts"},
+                exclude_none=True,
+            )
+
+            for typ_field, typ_value in brand_typography.items():
+                if typ_field not in typography_map:
+                    if raise_for_unmapped_vars:
+                        raise ThemeBrandUnmappedFieldError(f"typography.{typ_field}")
+                    continue
+
+                for typ_field_key, typ_field_value in typ_value.items():
+                    if typ_field_key in typography_map[typ_field]:
+                        typo_sass_var = typography_map[typ_field][typ_field_key]
+
+                        if typ_field == "base" and typ_field_key == "size":
+                            typ_field_value = str(
+                                maybe_convert_font_size_to_rem(typ_field_value)
+                            )
+
+                        sass_vars_typography[typo_sass_var] = typ_field_value
+                    elif raise_for_unmapped_vars:
+                        raise ThemeBrandUnmappedFieldError(
+                            f"typography.{typ_field}.{typ_field_key}"
+                        )
```

Comment:
It'd be easier to get a feel for what `__init__` was doing if these each "chunk" of logic was pulled out into a helper method. Something kind of like:

```python
class ThemeBrand(Theme):
    def __init__(
        self,
        brand: Brand,
        *,
        include_paths: Optional[str | Path | list[str | Path]] = None,
        strict_mode: bool = False
    ):
        self.strict_mode = strict_mode
        self.config = BrandBootstrapConfig.from_brand(brand)
        
        name = self._get_theme_name(brand)
        super().__init__(
            name=name,
            preset=self.config.preset,
            include_paths=include_paths,
        )
        
        self._process_colors(brand)
        self._process_typography(brand)
        self._add_default_rules()
    
    def _process_colors(self, brand: Brand) -> None:
        if not brand.color:
            return
            
        color_vars = self._map_theme_colors(brand.color)
        palette_vars = self._map_palette_colors(brand.color)
        
        self.add_defaults(**color_vars, **palette_vars)
        
    def _map_theme_colors(self, colors: Any) -> dict[str, str]:
        """Map theme colors to Bootstrap variables"""
        mapped = {}
        for theme_name, color in colors.to_dict(include="theme").items():
            if theme_name not in color_map:
                self._handle_unmapped_field(f"color.{theme_name}")
                continue
                
            for sass_var in color_map[theme_name]:
                mapped[sass_var] = color
        return mapped
    
    @staticmethod
    def _sanitize_var_name(name: str) -> str:
        """Sanitize variable names for Sass/CSS use"""
        name = re.sub(r"""['"]""", "", name)
        return re.sub(r"[^a-zA-Z0-9_-]+", "-", name.lower())
    
    def _handle_unmapped_field(self, field: str) -> None:
        """Handle unmapped brand.yml fields based on strict mode"""
        if self.strict_mode:
            raise ThemeBrandUnmappedFieldError(field)
```

Reply:
Agreed, I had the same thought but it's been helpful to have it all in one method while working on it. Now that things are stable, I'll do this or something like it

Reply:
I refactored this and broke it up into smaller methods. One big difference from the above suggestion is that I did not try to have the methods that process color/typography also interact with the Sass layers. The ordering is important and it's easy to get wrong, especially with the defaults where each call is _prepended_ rather than _appended_.

Reply:
We talked through this in person and refactored again together.